test/utils/string_manipulation_test.exs

```diff
@@ -0,0 +1,53 @@
+defmodule JSONAPI.Utils.StringManipulationTest do
+  @moduledoc false
+
+  use ExUnit.Case, async: true
+
+  import JSONAPI.Utils.StringManipulation
+
+  doctest JSONAPI.Utils.StringManipulation
+
+  describe "maybe_dasherize/1 when :underscore_to_dash is true" do
+    setup do
+      Application.put_env(:jsonapi, :underscore_to_dash, true)
+
+      on_exit(fn ->
+        Application.delete_env(:jsonapi, :underscore_to_dash)
+      end)
+
+      {:ok, []}
+    end
+
+    test "dasherizes a term" do
+      assert maybe_dasherize("foo_bar") == "foo-bar"
+    end
+  end
+
+  describe "maybe_dasherize/1 when :underscore_to_dash is false" do
+    test "does not dasherize a term" do
+      assert maybe_dasherize("foo_bar") == "foo_bar"
+    end
+  end
+
+  describe "maybe_underscore/1 when :underscore_to_dash is true" do
+    setup do
+      Application.put_env(:jsonapi, :underscore_to_dash, true)
+
+      on_exit(fn ->
+        Application.delete_env(:jsonapi, :underscore_to_dash)
+      end)
+
+      {:ok, []}
+    end
+
+    test "underscores a term" do
+      assert maybe_underscore("foo-bar") == "foo_bar"
```

Comment:
So if config says keys should be dasherized, then underscore?  Same behaviour on master, but confusing 😆

Reply:
Should be better now. I've left room for the new option to be used for camelizing.